### DISCUSSION TITLE
deployment issues with net48 webapps using SDK style project files?

### DISCUSSION THREAD

**Discussion by palmej2** [CZEMacLeod/MSBuild.SDK.SystemWeb#81] [General]
Hello - I have a few different ASP.NET MVC projects that build on .NET Framework 4.8 I want to convert to the new SDK style csproj format. I was able to convert it over fine, but once I try and publish the project, all of the DLL files are not located in a /bin folder and nothing works. Instead they are just all dumped in one big folder with the other web assets (CSS, JS, etc) are at the same level. 

Question: Will using this repo solve the deployment problems with net48 webapps?  I tried to use the <Project Sdk="MSBuild.SDK.SystemWeb/4.0.88"> in my local solution but hit various problems;

-on publish i got this error: error ASPCONFIG: The CodeDom provider type "Microsoft.VisualC.CppCodeProvider, CppCodeProvider, Version=10.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a" could not be located.

-on publish it went into a endless loop with this error: ASPNETCOMPILER : warning CS1685: The predefined type 'System.ObsoleteAttribute' is defined in multiple assemblies in the global alias; using definition from 'c:\Program Files (x86)\Reference Assemblies\Microsoft\Framework\.NETFramework\v4.8\mscorlib.dll'

For what it's worth my only issue with the standard MS SDK on a .NET 48 webapp was the deployment issues as well.  I had conversation with them [here](https://learn.microsoft.com/en-us/answers/questions/2074958/converting-asp-net-mvc-framework-csproj-question?comment=question#comment-1798743). 

UPDATE: I did try the using the Web Application (Model-View-Controller) (CZEMacLeod)" template in Visual Studio and here are the deployment / publish issues I faced:

- trying to use the standard publish in Visual Studio for that project, it said it published correctly but the publish folder was blank? 
- then i tried using dotnet publish ./webapplication3.csproj -c Release -o C:\custom\deploymentFolder and it failed with other errors
 -   C:\Users[my user id].nuget\packages\msbuild.sdk.systemweb\4.0.88\Sdk\Sdk.targets(31,3): error MSB4019: The imported project "C:\Program Files\dotnet\sdk\8.0.401\Microsoft\VisualStudio\v17.0\WebApplications\Microsoft.WebApplication.targets" was not found. Co nfirm that the expression in the Import declaration "C:\Program Files\dotnet\sdk\8.0.401\Microsoft\VisualStudio\v17.0\WebApplications\Microsoft.WebApplication.targets" is correct, and that the file exists on disk. [C:\Users[my user id]\source\repos\WebApplic ation3\webapplication3.csproj]

Thanks

**Comment by CZEMacLeod** [comment]
Hi most of this is covered in #12 and is one of the main limitations.
While this uses SDK style projects, it is still net framework dependent and you must use `msbuild`, not `dotnet` to build and publish as it requires the full framework for some tasks.
There are also some notes in https://github.com/CZEMacLeod/MSBuild.SDK.SystemWeb/issues/21#issuecomment-1478474537

The basic upshot is that you can publish from the command line or CI exactly as you could previously, but the IDE experience doesn't behave well.
As long as you are not doing 'green field' development, you probably have all your 'head' projects already and have the publish profiles setup.
You can bring these across as needed and use something like
```cmd
msbuild /t:Publish /p:DeployOnBuild=true /p:WebPublishMethod=Package /p:PackageAsSingeFile=true /p:PackageLocation="MyPublishDir"
```
or
```cmd
msbuild "path\to\your\solution.sln" /t:rebuild ^
				    /p:Configuration=Release ^
				    /p:DeployOnBuild=True ^
				    /p:DeployDefaultTarget=WebPublish ^
				    /p:WebPublishMethod=FileSystem ^
				    /p:DeleteExistingFiles=True ^
				    /p:publishUrl="path\to\your\output\folder"
```

Looking at your discussion, all of the copytooutput type stuff should not be needed using this SDK.

**Reply by palmej2** [reply]
@CZEMacLeod - thank you for your input!  So using your SDK ("MSBuild.SDK.SystemWeb/4.0.88") in my project file (instead of the standard "Microsoft.NET.Sdk.Web") along with the standard msdeploy scripts, everything works correctly (builds + deploys).  

Thanks!

**Comment by blmille1** [comment]
When will `dotnet build/publish` be supported?
Unfortunately, I tried to convert to this today and realized because of this thread that dotnet CLI is not supported.
I've made many changes to my repo to make it work with dotnet build and publish and I can't seem to get it to work again with msbuild.

I converted to SDK only because Microsoft said to migrate to SDK-style projects first and didn't warn me not to do it for System.Web-based projects.  It's been a month now and my CI/CD system still isn't fixed.  This is only a stepping stone for me to get to .NET.

**Reply by CZEMacLeod** [reply]
dotnet build is unlikely to ever be supported as that uses net core to do the build, and building asp.net projects requires the full framework which is why you need to use msbuild.
dotnet publish again is designed to publish net core / net 5+ projects, and does not support the web publishing methods used in asp.net 4.x.
CI/CD systems can use msbuild to build the projects successfully as mentioned above.
You can use msbuild to build net8/9 projects so if you are using systemweb-adapters project to do a cutover migration, then you can use that until you remove the asp.net project.
